tests/test_app_fs.py

```diff
@@ -0,0 +1,280 @@
+import pathlib
+import shutil
+import tempfile
+from typing import Optional
+from unittest.mock import MagicMock, Mock, call, patch
+
+import pytest
+
+
+@pytest.fixture
+def mock_copy2():
+    # mock pi_top_usb_setup.app_fs.copy2
+    with patch("pi_top_usb_setup.app_fs.copy2") as copy_mock:
+        yield copy_mock
+
+
+@pytest.fixture
+def mock_makedirs():
+    with patch("pi_top_usb_setup.app_fs.makedirs") as makedirs_mock:
+        yield makedirs_mock
+
+
+@pytest.fixture
+def mock_run_command(mock_pitop_imports):
+    command_runner_mock = mock_pitop_imports["pitop.common.command_runner"]
+    command_runner_mock.run_command = MagicMock()
+    yield command_runner_mock.run_command
+
+
+def create_structure(structure: dict, base_path: pathlib.Path):
+    """
+    Recursively create files and folders from a nested dictionary structure.
+
+    Args:
+        structure (dict): Dictionary representing folder/file structure.
+        base_path (Path): Base directory to start creating the structure.
+    """
+    for name, content in structure.items():
+        current_path = base_path / name
+        if isinstance(content, dict):
+            # It's a folder
+            current_path.mkdir(parents=True, exist_ok=True)
+            create_structure(content, current_path)
+        else:
+            # It's a file with optional content
+            current_path.touch()
+            if content:
+                current_path.write_text(content)
+
+
+@pytest.fixture
+def create_mount_point():
+    def _create_mount_point(files: dict):
+        temp_dir = tempfile.mktemp()
+        mount_point = pathlib.Path(temp_dir)
+        mount_point.mkdir(exist_ok=True)
+
+        create_structure(files, mount_point)
+        return temp_dir
+
+    yield _create_mount_point
+
+
+@pytest.fixture
+def app_fs(create_mount_point):
+    created_dirs = []
+
+    def _create_app_fs(files: Optional[dict] = None):
+
+        temp_dir = create_mount_point(files)
+        created_dirs.append(temp_dir)
+
+        from pi_top_usb_setup.app_fs import AppFilesystem
+
+        AppFilesystem.TEMP_FOLDER = temp_dir
+        return AppFilesystem(temp_dir)
+
+    yield _create_app_fs
+
+    # Cleanup all created directories
+    for temp_dir in created_dirs:
+        shutil.rmtree(temp_dir)
+
+
+def test_copy_files_creates_file_structure_in_target(mock_copy2, mock_makedirs, app_fs):
+    # Test basic copy
+    structure = {
+        "pi-top-usb-setup.tar.gz": "",
+        "pi-top-usb-setup": {
+            "files": {
+                "tmp": {
+                    "folder": {
+                        "file.txt": "Hello, world!",
+                    },
+                    "file2.txt": "bla",
+                },
+            },
+        },
+    }
+    app = app_fs(structure)
+    app.copy_files()
+
+    base_path = app.FOLDER_TO_COPY
+
+    # Directories were created
+    mock_makedirs.assert_called_with("/tmp/folder", exist_ok=True)
```

Comment:
this should be called twice so let's add the other one too:
mock_makedirs.assert_called_with("/tmp", exist_ok=True)